src/main/kotlin/org/rust/ide/intentions/ImportNameIntention.kt

```diff
@@ -0,0 +1,232 @@
+            val pathMod = path.containingMod
+            val pathSuperMods = HashSet(pathMod.superMods)
+            // TODO: support reexports
+            val candidates = RsNamedElementIndex.findElementsByName(project, basePath.referenceName)
+                .filterIsInstance<RsQualifiedNamedElement>()
+                .filter { (it as? RsVisible)?.isVisibleIn(pathSuperMods) ?: false }
+            if (candidates.isNotEmpty()) {
+                return Context(path, candidates)
+            }
+        }
+
+        return null
+    }
+
+    // Semantic signature of method is `RsVisible.isVisibleIn(mod: RsMod)`
+    // but in our case `mod` is always same and `mod` needs only to get set of its super mods
+    // so we pass `superMods` instead of `mod` for optimization
+    private fun RsVisible.isVisibleIn(superMods: Set<RsMod>): Boolean {
+        val parentMod = (if (this is RsMod) `super` else containingMod) ?: return false
+        val ourSuperMods = parentMod.superMods
+
+        // try to find latest common ancestor module of `parentMod` and `mod` in module tree
+        // we need to do it because we can use direct child items of any super mod with any visibility
+        val lca = ourSuperMods.find { it in superMods }
+
+        val shouldBePublicMods = if (lca == null) {
+            if (!isPublic) return false
+            ourSuperMods
+        } else {
+            // if current item is direct child of some ancestor of `mod` then it can be not public
+            if (parentMod == lca) return true
+            if (!isPublic) return false
+            ourSuperMods.takeWhile { it != lca }.dropLast(1)
+        }
+        return shouldBePublicMods.all { it.isPublic }
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val (path, candidates) = ctx
+
+        if (candidates.size == 1) {
+            importItem(project, candidates.first(), path)
+        } else {
+            chooseItemAndImport(project, editor, candidates, path)
+        }
+    }
+
+    private fun chooseItemAndImport(project: Project, editor: Editor, items: List<RsQualifiedNamedElement>, originalPath: RsPath) {
+        // TODO: sort items in popup
+        val step = object : BaseListPopupStep<RsQualifiedNamedElement>("Item to Import", items) {
+            override fun isAutoSelectionEnabled(): Boolean = false
+            override fun isSpeedSearchEnabled(): Boolean = true
+            override fun hasSubstep(selectedValue: RsQualifiedNamedElement?): Boolean = false
+
+            override fun onChosen(selectedValue: RsQualifiedNamedElement?, finalChoice: Boolean): PopupStep<*>? {
+                if (selectedValue == null) return PopupStep.FINAL_CHOICE
+                return doFinalStep { importItem(project, selectedValue, originalPath) }
+            }
+
+            override fun getTextFor(value: RsQualifiedNamedElement): String =
+                ObjectUtils.assertNotNull(value.qualifiedName)
```

Comment:
Just `value.qualifiedName!!`?

Reply:
Yeah, `ObjectUtils.assertNotNull` is used in similar java import action and remained here by accident